src/sctools/metrics/metricsrunner.py

```diff
@@ -0,0 +1,61 @@
+import argparse
+import pysam
+
+
+class MetricsRunner:
+
+    def __init__(self, sam_file, open_mode=None):
+        """
+        :param str alignment_file: sam or bam file.
+        :param str open_mode: optional, mode to read file. Will be autodetected by file type if
+          the file contains the correct suffix for its type.
+        """
+
+        if open_mode is None:
+            if sam_file.endswith('.bam'):
+                open_mode = 'rb'
+            elif sam_file.endswith('.sam'):
+                open_mode = 'r'
+            else:
+                raise ValueError('could not autodetect file type for alignment_file %s '
+                                 '(detectible suffixes: .sam, .bam)' % sam_file)
+        self.input_sam = sam_file
+        self.open_mode = open_mode
+
+    def run_metrics(self, metrics_to_run):
+        """
+        :param list metrics_to_run: list of metrics to run for each record in input_sam
+        """
+
+        for metric in metrics_to_run:
+            metric.setup()
+
+        with pysam.AlignmentFile(self.input_sam, self.open_mode) as sam:
+            for record in sam:
+                for metric in metrics_to_run:
+                    metric.gather_metric(record)
+
+        for metric in metrics_to_run:
+            metric.calculate_and_output()
```

Comment:
I think we might want to split up `calculate` and `output`. I think the pattern as we've discussed is to have each metric class broken up by what its inputs are, but the _outputs_ might be split differently. I could imagine everything that produces a vector summarizing all the genes to want to go into the same output matrix, for example. 

Maybe some kind of general `output` that takes a bunch of calculated metrics in a particular shape and outputs them together? 

Thinking about memory usage, if those `calculate` commands generate large data structures that need to be dumped immediately to disk, then `output` could open up a file object and take a bunch of metrics as parameters, which it would then `calculate`, dump to the file, and grab the next one. 

... or is this overthinking because we could just concatenate new rows to a .csv.gz file? In any case, I want some kind of thoughtful solution here. :)

Reply:
The idea I had was that each metric would be given some kind of basename metric name and then each metric would output however many files it needs to given the basename.  I was thinking `.tsv` or `.csv` should be good memory friendly enough.

I think this should still just be one method because at this point our memory footprint would be at its largest and we just want to go through each metric and just calculate whatever we need to and have it write out the files.  Once its written its files all of its members could then be cleaned up as we move on and do the same to the next metric.  I wouldn't want to `calculate` on every metric making out memory footprint larger than it has to be.  I definitely think we should think this through though.

Reply:
Not 100% sure what you mean by `basename` here. We definitely will want any comparable metrics to sit in the same file so they can be accessed via PCA or something similar. 

That said, I think the latter part of my ramblings above identifies that we can just open up a file in append mode and dump another row into it, which accomplishes both our requirements.